Draft!: Broken into many PRs created cogroup option

wcarlson5 opened the pull request:
Created the option to make a KCogroupStream object from a grouped stream. The KCogroupStream object then can be further cogrouped or aggregated.

This PR has been broken into parts for review 
first part: https://github.com/apache/kafka/pull/7538

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
first commit on Cogroup operator

wcarlson5 committed:
first commit on Cogroup operator

wcarlson5 committed:
Added some javadocs and unit tests

wcarlson5 commented:
@mjsax definitely not done but I think I am almost ready to start working on the windowed stores.

wcarlson5 committed:
refactored for window stores

wcarlson5 committed:
fixed style

wcarlson5 committed:
Added TimeWindow option for cogroup and made basic tests + expaned tests for KV cogroup

wcarlson5 committed:
Expanded tests for TimeWindowedKCogroupedStreams

wcarlson5 committed:
added SessionWindows and tests

wcarlson5 committed:
updated javadocs and style

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KCogroupedStream.java:
```diff
@@ -0,0 +1,194 @@
+ * @param <K> Type of keys
+ * @param <V> Type of values
+ */
+public interface KCogroupedStream<K, T, V> {
```
I think this should be `CogroupedKStream`.

mjsax replied:
We should update the generic types `T -> VIn` and `V -> VOut` ?

wcarlson5 replied:
> I think this should be `CogroupedKStream`.

For GroupedStreams it is KGroupedStream. I am ambivalent but am not sure we should deviate from the current pattern

wcarlson5 replied:
T and V gave me such a head ache. I am defiantly renaming them

mjsax replied:
`KGroupedStream` is the old naming pattern. Cf. `TimeWindowedKStream` or `SessionWindowedKStream` that apply the new pattern already.

wcarlson5 committed:
updated javadocs and style

wcarlson5 committed:
Merge pull request #1 from apache/trunk

update my fork

wcarlson5 committed:
make it so I can use the drive for testing
Merge branch 'trunk' of github.com:wcarlson5/kafka into KAFKA-6049_created_cogroup_option

wcarlson5 committed:
updated tests to use testInputTopic

wcarlson5 committed:
small style changes

wcarlson5 committed:
added documentation

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+        this.builder = builder;
+    }
+
+    <KR, T, W extends Window> KTable<KR, V> build(
```
KR is required because for windowed tables the key needs to be be type Windowed<K>

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowTableSource.java:
```diff
@@ -0,0 +1,48 @@
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+
+class KStreamWindowTableSource<K, V> implements ProcessorSupplier<K, V> {
```
This class is needed because there is a table source that can take input form many different processors but it only worked with KeyValue records. The table constructor used to be able to take multiple nodes but it no longer does So the different streams need to be consolidated first.

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedKCogroupedStreamImpl.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    @SuppressWarnings("deprecation") // continuing to support SessionWindows#maintainMs in fallback mode
+    private  StoreBuilder<SessionStore<K, V>> materialize(final MaterializedInternal<K, V, SessionStore<Bytes, byte[]>> materialized) {
```
This is following the pattern in the non cogrouped version.

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKCogroupedStreamImpl.java:
```diff
@@ -0,0 +1,167 @@
+import org.apache.kafka.streams.state.WindowStore;
+import org.apache.kafka.streams.state.internals.RocksDbWindowBytesStoreSupplier;
+
+public class TimeWindowedKCogroupedStreamImpl<K, T, V, W extends Window> extends
```
This is following the pattern in the non cogrouped version.

mjsax replied:
rename -> `TimeWindowedCogroupedKStreamImpl` (similar the interface)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KCogroupedStream.java:
```diff
@@ -0,0 +1,194 @@
+ * KGroupedStream#cogroup(Aggregator, Materialized) cogroup(...)}.
+ *
+ * @param <K> Type of keys
+ * @param <V> Type of values
```
what "values"

Also `T` is missing

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KCogroupedStream.java:
```diff
@@ -0,0 +1,194 @@
+     * represent the latest (rolling) aggregate for each key
+     */
+    KTable<K, V> aggregate(final Initializer<V> initializer,
+                           final Materialized<K, V, KeyValueStore<Bytes, byte[]>> materialized);
```
Should we have one overload that only takes a single argument `Initializer` ?

wcarlson5 replied:
achieved by removing StoreSupplier

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KCogroupedStream.java:
```diff
@@ -0,0 +1,194 @@
+     * represent the latest (rolling) aggregate for each key
+     */
+    KTable<K, V> aggregate(final Initializer<V> initializer,
+                           final StoreSupplier<KeyValueStore> storeSupplier);
```
We don't expose `StoreSupplier` in the DSL any longer -- I think this overload can be removed.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -42,201 +42,202 @@
-     * Note that the internal store name may not be queriable through Interactive Queries.
-     *
+     * Count the number of records in this stream by the grouped key. Records with {@code null} key
+     * or value are ignored. The result is written into a local {@link KeyValueStore} (which is
```
Please update you IDE setting to avoid unnecessary reformatting.

We also apply a "start a new line for each new sentence" policy, because it minimizes PR diffs if we update the JavaDocs and thus makes reviewing easier.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -309,230 +311,275 @@
+     *
+     * @param aggregator an {@link Aggregator} that computes a new aggregate result
+     */
+    <T> KCogroupedStream<K, V, T> cogroup(final Aggregator<? super K, ? super V, T> aggregator);
```
We should update the generic type -- maybe you can to small MINOR PR that fixed generic type naming for all classes (we often use `V`, `V2`, `VR` etc and those are all terrible -- would be worth a small cleanup PR on the side.) (Cf `KStreamFlatTransform` that uses the new/better naming schema already).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -309,230 +311,275 @@
+     * aggregations are applied to the new partitions resulting in a {@link KTable}.
+     * <p>
+     *
+     * @param aggregator an {@link Aggregator} that computes a new aggregate result
```
Generic type `T` should be documented

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+
+class CogroupedStreamAggregateBuilder<K, V> {
+
+    private static final String AGGREGATE_NAME = "KCOGROUPSTREAM-AGGREGATE-";
```
nit: `COGROUPKSTREAM-AGGREGATE -`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+
+    private <T, W extends Window> StatefulProcessorNode getStatefulProcessorNode(
+        final Aggregator<? super K, ? super T, V> aggregator,
+        final Initializer<V> initializer, final NamedInternal named, final boolean stateCreated,
```
nit: formatting -> only one parameter per line

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+        if (windows == null && sessionWindows == null) {
+            kStreamAggregate = new KStreamAggregate<>(storeBuilder.name(),
+                                                             initializer,
+                                                             aggregator);
```
nit: fix indention

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImpl.java:
```diff
@@ -0,0 +1,130 @@
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StoreSupplier;
+
+public class KCogroupedStreamImpl<K, T, V> extends AbstractStream<K, V> implements
```
rename `CogroupedKStreamImpl`

wcarlson5 replied:
it was at 100 , I moved it up to 100

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImpl.java:
```diff
@@ -0,0 +1,130 @@
+import org.apache.kafka.streams.state.StoreSupplier;
+
+public class KCogroupedStreamImpl<K, T, V> extends AbstractStream<K, V> implements
+    KCogroupedStream<K, T, V> {
```
nit: seem this should go in line above -- seems you IDE setting have a too narrow max line length setting?

wcarlson5 replied:
it was at 100 , I moved it up to 100

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedKCogroupedStreamImpl.java:
```diff
@@ -0,0 +1,136 @@
+                                          != null
+                                          ? new WindowedSerdes.SessionWindowedSerde<>(
+                                          materializedInternal
+                                              .keySerde()) : null,
```
nit formatting (hard to read)
```
materializedInternal.keySerde() != null ?
    new WindowedSerdes.SessionWindowedSerde<>(
        materializedInternal.keySerde()) :
    null,
```

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedKCogroupedStreamImpl.java:
```diff
@@ -0,0 +1,136 @@
+        if (supplier == null) {
+            // NOTE: in the future, when we remove sessionWindows#maintainMs(), we should set the default retention
+            // to be (sessionWindows.inactivityGap() + sessionWindows.grace()). This will yield the same default behavior.
+            final long retentionPeriod = materialized.retention() != null ? materialized.retention().toMillis() : sessionWindows.maintainMs();
```
Wondering if we can get rid of supporting deprecated API for this case? \cc @vvcephei WDYT?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+    @Test
+    public void cogroupTest() {
+        assertNotNull(cogroupedStream);
+    }
```
Think we can remove this test

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullKGroupedStreamOnCogroup() {
```
nit: `shouldThrowNPEInCogroupIfKGroupedStreamIsNull` (use descriptive name: "should not have" is not very precise)

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+        cogroupedStream.windowedBy((Windows) null);
+    }
+
+    private static final Aggregator STRING_AGGREGATOR = (Aggregator<String, String, String>) (key, value, aggregate) ->
```
nit: move all uses variable to the top of the class

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+    };
+
+    @Test
+    public void testCogroupBassicOneTopic() {
```
`shouldCoGroupAndAggregateSingleKStreams`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+
+        customers.toStream().to("to-one", Produced.with(Serdes.String(), Serdes.String()));
+
+        builder.stream("to-one", stringConsumed).process(processorSupplier);
```
Why do we need this? Why not use `driver.createOutputTopic(...)` to verify the output?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+
+        try (final TopologyTestDriver driver = new TopologyTestDriver(builder.build(), props)) {
+            final TestInputTopic<String, String> testInputTopic = driver.createInputTopic("one", new StringSerializer(), new StringSerializer());
+            testInputTopic.pipeInput("1", "A", 0);
```
nit: `1` -> `k1` to make clear it's the key

Shy do we need to set the timestamp explicitly?

wcarlson5 replied:
fixed key, we don't technically need the time stamp in all tests but I don't think it hurts

mjsax replied:
IMHO, it's distracting as if I read the test a start to reason about the timestamps. If it does not matter we should just use the auto-assigned timestamp to avoid "noise" in the test code.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+
+
+    @Test
+    public void testCogroupEachTopicUnique() {
```
`shouldCogroupAndAggreagetTwoKStreams`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+        final KGroupedStream groupedOne = test.groupByKey();
+
+        final KTable customers = groupedOne.cogroup(STRING_AGGREGATOR, Materialized.as("store"))
+            .aggregate(STRING_INITIALIZER, Materialized.as("store1"));
```
do we need to pass in store names? Why do we specify two stores with the same name?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+
+        final KTable customers = groupedOne.cogroup(STRING_AGGREGATOR, Materialized.as("store"))
+            .cogroup(groupedTwo, STRING_AGGREGATOR)
+            .aggregate(STRING_INITIALIZER, Materialized.as("store1"));
```
as above?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+
+        customers.toStream().to("to-one", Produced.with(Serdes.String(), Serdes.String()));
+
+        builder.stream("to-one", stringConsumed).process(processorSupplier);
```
as above

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+    @Test
+    public void testCogroupBassicOneTopic() {
+
+        final KStream test = builder.stream("one", stringConsumed);
```
Add generic types (similar for all other classes) -- I am sure IDE should warn you about raw types (if not, enable those warnings)

wcarlson5 replied:
there is one case where we cannot but added everywhere else

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+
+
+    @Test
+    public void testCogroupKeyMixedInTopics() {
```
What is the purpose of this test? Seems to be the same as above?

wcarlson5 replied:
the keys are mixed across topics where in the other test the key sets are disjoint

mjsax replied:
I see -- do we need both tests? Seem that this test super seed the other one?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowTableSource.java:
```diff
@@ -0,0 +1,48 @@
+
+    private class KStreamJoinWindowProcessor extends AbstractProcessor<K, V> {
+
+        @SuppressWarnings("unchecked")
```
why do need this suppression?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KCogroupedStreamImplTest.java:
```diff
@@ -0,0 +1,403 @@
+import org.junit.Before;
+import org.junit.Test;
+
+@SuppressWarnings("unchecked")
```
We should not suppress warning globally but rather avoid them (if the cannot be avoided, suppress on method level only)

wcarlson5 replied:
some of the type conversions when dealing with GroupedStreams with different value type can not be checked. I have made it specific to them only

mjsax reviewed (commented):
Did an initial high level pass over it.

wcarlson5 committed:
responded to 90% of the comments

wcarlson5 committed:
reformmated javadocs

wcarlson5 committed:
made input a ?

wcarlson5 committed:
made input a ? and removed option to set materialized in cogroup

wcarlson5 committed:
Merge pull request #2 from apache/trunk

update my fork

wcarlson5 committed:
Merge branch 'trunk' into KAFKA-6049_created_cogroup_option

wcarlson5 committed:
made input a ? and removed option to set materialized in cogroup

wcarlson5 committed:
enforced copartitioning for cogroups

wcarlson5 committed:
removed duplicate import

wcarlson5 committed:
resolved typing errors

wcarlson5 committed:
resolved typing errors

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -0,0 +1,125 @@
+        this.aggregateBuilder = new CogroupedStreamAggregateBuilder<>(builder);
+    }
+
+    @SuppressWarnings("unchecked")
```
in order for processor suppliers to receive an aggregator it must have a typed input value that extends V.  We solve this by making V object and casting to V but this needs a suppression.

mjsax commented:
Closing this draft PR -- KIP-150 is completed already.
